graphix/fundamentals.py

```diff
@@ -0,0 +1,308 @@
+        """Return `1.0` for `+` and `-1.0` for `-`."""
+        return float(self.value)
+
+    def __complex__(self) -> complex:
+        """Return `1.0 + 0j` for `+` and `-1.0 + 0j` for `-`."""
+        return complex(self.value)
+
+
+class ComplexUnit(Enum):
+    """
+    Complex unit: 1, -1, j, -j.
+
+    Complex units can be multiplied with other complex units,
+    with Python constants 1, -1, 1j, -1j, and can be negated.
+    """
+
+    # HACK: complex(u) == (1j) ** u.value for all u in ComplexUnit.
+
+    ONE = 0
+    J = 1
+    MINUS_ONE = 2
+    MINUS_J = 3
+
+    @staticmethod
+    def try_from(value: ComplexUnit | SupportsComplexCtor) -> ComplexUnit | None:
+        """Return the ComplexUnit instance if the value is compatible, None otherwise."""
+        if isinstance(value, ComplexUnit):
+            return value
+        try:
+            value = complex(value)
+        except Exception:
+            return None
+        if value == 1:
+            return ComplexUnit.ONE
+        if value == -1:
+            return ComplexUnit.MINUS_ONE
+        if value == 1j:
+            return ComplexUnit.J
+        if value == -1j:
+            return ComplexUnit.MINUS_J
+        return None
+
+    @staticmethod
+    def from_properties(*, sign: Sign = Sign.PLUS, is_imag: bool = False) -> ComplexUnit:
+        """Construct ComplexUnit from its properties."""
+        osign = 0 if sign == Sign.PLUS else 2
+        oimag = 1 if is_imag else 0
+        return ComplexUnit(osign + oimag)
+
+    @property
+    def sign(self) -> Sign:
+        """Return the sign."""
+        return Sign.plus_if(self.value < 2)
+
+    @property
+    def is_imag(self) -> bool:
+        """Return `True` if `j` or `-j`."""
+        return bool(self.value % 2)
+
+    def __complex__(self) -> complex:
+        """Return the unit as complex number."""
+        ret: complex = 1j**self.value
+        return ret
+
+    def __str__(self) -> str:
+        """Return a string representation of the unit."""
+        result = "1j" if self.is_imag else "1"
+        if self.sign == Sign.MINUS:
+            result = "-" + result
+        return result
+
+    def __mul__(self, other: ComplexUnit | SupportsComplexCtor) -> ComplexUnit:
+        """Multiply the complex unit with a number."""
+        if isinstance(other, ComplexUnit):
+            return ComplexUnit((self.value + other.value) % 4)
+        if other_ := ComplexUnit.try_from(other):
+            return self.__mul__(other_)
+        return NotImplemented
+
+    def __rmul__(self, other: SupportsComplexCtor) -> ComplexUnit:
+        """Multiply the complex unit with a number."""
+        return self.__mul__(other)
+
+    def __neg__(self) -> ComplexUnit:
+        """Return the opposite of the complex unit."""
+        return ComplexUnit((self.value + 2) % 4)
+
+
+class IXYZ(Enum):
+    """I, X, Y or Z."""
+
+    I = enum.auto()
+    X = enum.auto()
+    Y = enum.auto()
+    Z = enum.auto()
+
+    @functools.cached_property
```

Comment:
There is no computation to cache in this method.

Reply:
Reverted in b40b7eafb4e74a0fac424deb96a778f2ce5d0cdb .